pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -0,0 +1,158 @@
+	configs := c.config.ExternalServices()
+
+	services := make([]*model.Service, 0)
+	for _, externalServiceConfig := range configs {
+		externalService := externalServiceConfig.Spec.(*networking.ExternalService)
+
+		services = append(services, convertService(externalService)...)
+	}
+
+	return services, nil
+}
+
+// GetService retrieves a service by host name if it exists
+func (c *externalDiscovery) GetService(hostname string) (*model.Service, error) {
+
+	// Get actual service by name
+	name, err := parseHostname(hostname)
+	if err != nil {
+		log.Infof("parseHostname(%s) => error %v", hostname, err)
+		return nil, err
+	}
+
+	for _, service := range c.getServices() {
+		if service.Hostname == name {
+			return service, nil
+		}
+	}
+
+
+	return nil, nil
+}
+
+func (c *externalDiscovery) getServices() ([]*model.Service) {
+	configs := c.config.ExternalServices()
+
+	services := make([]*model.Service, 0)
+	for _, externalServiceConfig := range configs {
+		externalService := externalServiceConfig.Spec.(*networking.ExternalService)
+
+		services = append(services, convertService(externalService)...)
+	}
+	return services
+}
+
+
+// ManagementPorts retries set of health check ports by instance IP.
+// This does not apply to Consul service registry, as Consul does not
+// manage the service instances. In future, when we integrate Nomad, we
+// might revisit this function.
+func (c *externalDiscovery) ManagementPorts(addr string) model.PortList {
+	return nil
+}
+
+// Instances retrieves instances for a service and its ports that match
+// any of the supplied labels. All instances match an empty tag list.
+func (c *externalDiscovery) Instances(hostname string, ports []string,
+	labels model.LabelsCollection) ([]*model.ServiceInstance, error) {
+	// Get actual service by name
+	name, err := parseHostname(hostname)
+	if err != nil {
+		log.Infof("parseHostname(%s) => error %v", hostname, err)
+		return nil, err
+	}
+
+	portMap := make(map[string]bool)
+	for _, port := range ports {
+		portMap[port] = true
+	}
+
+	instances := []*model.ServiceInstance{}
+	externalInstances := []*model.ServiceInstance{}
+
+	configs := c.config.ExternalServices()
+	for _, externalServiceConfig := range configs {
+		externalService := externalServiceConfig.Spec.(*networking.ExternalService)
+		externalInstances = append(externalInstances, convertInstances(externalService)...)
+	}
+
+	for _, externalInstance := range externalInstances {
+		if externalInstance.Service.Hostname == name &&
+			labels.HasSubsetOf(externalInstance.Labels) &&
+			portMatch(externalInstance, portMap) {
+
+			instances = append(instances, externalInstance)
+		}
+	}
+
+	return instances, nil
+}
+
+// returns true if an instance's port matches with any in the provided list
+func portMatch(instance *model.ServiceInstance, portMap map[string]bool) bool {
+	if len(portMap) == 0 {
+		return true
+	}
+
+	if portMap[instance.Endpoint.ServicePort.Name] {
+		return true
+	}
+
+	return false
+}
```

Comment:
could be a single line:
`return len(portMap) == 0 || portMap[instance.Endpoint.ServicePort.Name]`